programming_examples/matrix_scalar_add/matrix_scalar_add_implicit.py

```diff
@@ -0,0 +1,88 @@
+# Copyright (C) 2024, Advanced Micro Devices, Inc.
```

Comment:
This example still feels a bit like magic to me. I can see that it somehow works - all the data in the image is updated, not just one tile - but it's opaque to me at what level and how some sort of looping behavior eventually is introduced.

I looked at the mlir generated by `python matrix_scalar_add_implicit.py`, and I get:
```mlir
module {
  air.channel @ChanIn []
  air.channel @ChanOut []
  func.func @addone(%arg0: memref<32x16xi32>, %arg1: memref<32x16xi32>) {
    air.launch () in () args(%arg2=%arg0, %arg3=%arg1) : memref<32x16xi32>, memref<32x16xi32> {
      air.channel.put  @ChanIn[] (%arg2[] [] []) : (memref<32x16xi32>)
      air.channel.get  @ChanOut[] (%arg3[] [] []) : (memref<32x16xi32>)
      air.segment @seg  {
        %c1 = arith.constant 1 : index
        %c1_0 = arith.constant 1 : index
        air.herd @addherd  tile (%arg4, %arg5) in (%arg6=%c1, %arg7=%c1_0) {
          %alloc = memref.alloc() : memref<16x8xi32, 2 : i32>
          %alloc_1 = memref.alloc() : memref<16x8xi32, 2 : i32>
          air.channel.get  @ChanIn[] (%alloc[] [] []) : (memref<16x8xi32, 2 : i32>)
          %c0 = arith.constant 0 : index
          %c8 = arith.constant 8 : index
          %c1_2 = arith.constant 1 : index
          scf.for %arg8 = %c0 to %c8 step %c1_2 {
            %c0_3 = arith.constant 0 : index
            %c16 = arith.constant 16 : index
            %c1_4 = arith.constant 1 : index
            scf.for %arg9 = %c0_3 to %c16 step %c1_4 {
              %0 = memref.load %alloc[%arg9, %arg8] : memref<16x8xi32, 2 : i32>
              %c1_i32 = arith.constant 1 : i32
              %1 = arith.addi %0, %c1_i32 : i32
              memref.store %1, %alloc_1[%arg9, %arg8] : memref<16x8xi32, 2 : i32>
            }
          }
          air.channel.put  @ChanOut[] (%alloc_1[] [] []) : (memref<16x8xi32, 2 : i32>)
          memref.dealloc %alloc : memref<16x8xi32, 2 : i32>
          memref.dealloc %alloc_1 : memref<16x8xi32, 2 : i32>
          air.herd_terminator
        }
        air.segment_terminator
      }
      air.launch_terminator
    }
    return
  }
}
```
I don't see the loop yet here either. I could go deeper, but as a sample "python AIR developer" I'm trying to see if I can use the python without being an expert in mlir.

I guess my question here is philosophical:
* Do we want to present this behavior as 'trust the compiler, you don't need to know what it's doing'
* Or do we want to say, 'you can do this, but the behavior is ambiguous so it's best to be specific at the python level instead'? (a which point maybe I'd remove this example from the PR entirely).

Reply:
There is a `while(1)` loop generated at the top level of every `aie.core` because of the `emit-while-loop` option getting set to `true` here: https://github.com/Xilinx/mlir-air/blob/main/python/air/compiler/aircc/main.py#L399 (if you are compiling via aircc, which you are if you're using the xrt wrapper). I agree it's confusing and I had it set to `false` until a few weeks ago.

The reason that I set it to `true` is for benchmarking and I assume it's the same reason there's a `for i = 0...int_max` in many of the mlir-aie examples. Our usual benchmarking host code will typically do some number of warmup iterations of the kernel followed by some number of timed iterations. The problem I had was that xrt did not reconfigure the aie array between calls of the same kernel. So, if I programmed the aie cores to do exactly one invocation worth of tiles, then each core would happily do so for the first invocation, halt, then everything would hang the second time around.

Reply:
Ah! Thanks for the explanation.

Would it make sense to disable ```emit-while-loop``` again and then add some sort of `--benchmarking` option in aircc that enables it on demand?

Reply:
The option should be exposed to the aircc user, but there are use cases other than benchmarking.